plutus-tx/src/Language/PlutusTx/Data.hs

```diff
@@ -51,50 +50,99 @@ https://github.com/well-typed/cborg/issues/222). So we need to be permissive
-viewList (CBOR.TList l)  = Just l
-viewList (CBOR.TListI l) = Just l
-viewList _               = Nothing
-
-viewMap :: CBOR.Term -> Maybe [(CBOR.Term, CBOR.Term)]
-viewMap (CBOR.TMap m)  = Just m
-viewMap (CBOR.TMapI m) = Just m
-viewMap _              = Nothing
-
-viewBytes :: CBOR.Term -> Maybe BS.ByteString
-viewBytes (CBOR.TBytes b)   = Just b
-viewBytes (CBOR.TBytesI lb) = Just $ BSL.toStrict lb
-viewBytes _                 = Nothing
-
-viewInteger :: CBOR.Term -> Maybe Integer
-viewInteger (CBOR.TInt i)     = Just (fromIntegral i)
-viewInteger (CBOR.TInteger i) = Just i
-viewInteger _                 = Nothing
-
--- TODO: try and make this match the serialization of Haskell types using derived 'Serialise'?
--- Would at least need to handle special cases with Null etc.
-fromTerm :: CBOR.Term -> Either CBORToDataError Data
-fromTerm = \case
-    -- See Note [Permissive decoding]
-    (viewInteger -> Just i) -> pure $ I i
-    (viewBytes -> Just b) -> pure $ B b
-    (viewMap -> Just entries) -> Map <$> traverse (bitraverse fromTerm fromTerm) entries
-    (viewList -> Just ts) -> List <$> traverse fromTerm ts
-    CBOR.TTagged i (viewList -> Just entries) -> Constr (fromIntegral i) <$> traverse fromTerm entries
-    t -> Left $ "Unsupported CBOR term: " ++ show t
-
-toTerm :: Data -> CBOR.Term
-toTerm = \case
-    I i -> CBOR.TInteger i
-    B b -> CBOR.TBytes b
-    Map entries -> CBOR.TMap (fmap (bimap toTerm toTerm) entries)
-    List ts -> CBOR.TList (fmap toTerm ts)
-    Constr i entries -> CBOR.TTagged (fromIntegral i) $ CBOR.TList $ fmap toTerm entries
+{- Note [Encoding of Data]
+
+All constructors of 'Data' map directly to CBOR primitives, *except* for the
+'Constr' constructor. So to encode a 'Data' value we first write a 'Bool'
+indicating whether the value is a 'Constr' or something else.
+
+When decoding 'Data', if the leading bit is true we know that we are looking at
+a 'Constr', so we can read the two arguments. If it is something else, we
+look at the CBOR token type.
+
+-}
+
+-- | Check whether this is using the 'Constr' constructor
+isConstr :: Data -> Bool
+isConstr = \case
+    Constr _ _ -> True
+    _ -> False
+
+encodeData :: Data -> Encoding
+encodeData dt = CBOR.encodeBool (isConstr dt) <> encodeRest where
+    encodeRest = case dt of
+            I i -> CBOR.encodeInteger i
+            B b -> CBOR.encodeBytes b
+            Map entries ->
+                CBOR.encodeMapLenIndef
+                    <> mconcat [ encodeData k <> encodeData v | (k, v) <- entries ]
+                    <> CBOR.encodeBreak
+            List ts ->
+                CBOR.encodeListLenIndef
+                    <> mconcat (encodeData <$> ts)
+                    <> CBOR.encodeBreak
+            Constr i entries ->
+                CBOR.encodeInteger i
+                    <> CBOR.encodeListLenIndef
+                    <> mconcat (encodeData <$> entries)
+                    <> CBOR.encodeBreak
+
+decodeData :: Decoder s Data
```

Comment:
This seems very low-level.  Can we not do the same sort of thing we do in `Language.PlutusCore.CBOR` and let cborg take care of encoding the basic types for us?  See in particular the part where it deals with `Term`: https://github.com/input-output-hk/plutus/blob/dc689024c48430b8a4727768b26872174a25d43a/plutus-core/src/Language/PlutusCore/CBOR.hs#L219

Reply:
good point! I'll change it

Reply:
I had to keep the low-level encoding for `Map`, otherwise it can't tell the difference between `[Data]` and `[(Data, Data)]`

Reply:
Looking at the plc serialisation code again, I see that we have a function called `encodeConcstructorTag` which just does the same as `encodeWord`;  I think it's just using a different name to make it clear what's going on.  Anyway, if you look here ...
https://github.com/input-output-hk/plutus/blob/dc689024c48430b8a4727768b26872174a25d43a/plutus-core/src/Language/PlutusCore/CBOR.hs#L58

... there's a comment saying "don't use `encodeTag`.  I think we used this in an earlier version and then Duncan told us not to because it was for something else.  That's what's used to encode `TTagged`, so I think we've fallen into exactly the same trap before!

Reply:
Can we just mimic what happens in plutus-core and say this?

```
encodeConstructorTag :: Word -> Encoding
encodeConstructorTag = encodeWord

decodeConstructorTag :: Decoder s Word
decodeConstructorTag = decodeWord

instance Serialise Data where
    encode = \case
             I i         -> encodeConstructorTag 0 <> encode i
             B b         -> encodeConstructorTag 1 <> encode b
             Map entries -> encodeConstructorTag 2 <> encode entries
             List ts     -> encodeConstructorTag 3 <> encode ts
             Constr n d  -> encodeConstructorTag 4 <> encode n <> encode d

    decode = go =<< decodeConstructorTag
        where go 0   = I <$> decode
              go 1   = B <$> decode
              go 2   = Map <$> decode
              go 3   = List <$> decode
              go 4   = Constr <$> decode <*> decode
              go tag = fail $ "Failed to decode Term: unexpected tag " ++ show tag
```

`cborg` seems to be able to magically work out what you want and do the right thing.  I think it'll automatically write an Integer using the most economical encoding, so it's probably not necessary to look for the individual integer typed when decoding. (?)

Reply:
But then we could just use the derived `Serialise` instance. The point of not using them was to use less space by not encoding all of the branches.

Reply:
Oh, I see.  The point is that there's not a lot of ambiguity in the `Data` type, so we can tell which constructor we've got just by looking at the data?  Do we know how much space that actually saves?

Reply:
Yes, that's right. I think the space saved is about 1 byte per constructor. Which is probably more relevant for the `Integer` s than for the `ByteString`s.

I thought about it some more, and another reason for doing what you suggested is that the encoding would be more stable (just like it is explained in Note [Stable encoding of PLC]). So I'm happy to make the change, but I'd like to know what @michaelpj thinks about it since he made the space-saving argument first

Reply:
I guess we could check to see what sort of difference it makes.  I also wonder how big the data scripts are in comparison with the other ones. Ah! I can find that out by serialising the extracted validators in the benchmarks.  Hold on...

Here's a handy table.  I've completely forgotten who pays for what by now.
```
crowdfunding/Context01.plc:    2719
crowdfunding/Context02.plc:    2719
crowdfunding/Context03.plc:    2719
crowdfunding/Context04.plc:    1661
crowdfunding/Context05.plc:    1661
crowdfunding/Datum01.plc:       359
crowdfunding/Datum02.plc:       359
crowdfunding/Datum03.plc:       359
crowdfunding/Redeemer01.plc:    333
crowdfunding/Redeemer02.plc:    333
crowdfunding/Validation.plc:  24206
crowdfunding/Validator01.plc: 20807

future/Context01.plc:    2177
future/Context02.plc:    3569
future/Context03.plc:    3569
future/Context04.plc:    2762
future/Context05.plc:    2448
future/Context06.plc:    2086
future/Context07.plc:    2448
future/Datum01.plc:       333
future/Datum02.plc:       359
future/Datum03.plc:       359
future/Datum04.plc:       668
future/Datum05.plc:       668
future/Redeemer01.plc:    333
future/Redeemer02.plc:    513
future/Redeemer03.plc:    700
future/Redeemer04.plc:    700
future/Validator01.plc: 13748
future/Validator02.plc: 21838
future/Validator03.plc: 42070

marlowe/trustfund/Context01.plc:     3683
marlowe/trustfund/Context02.plc:     1686
marlowe/trustfund/Datum01.plc:       2046
marlowe/trustfund/Datum02.plc:       1551
marlowe/trustfund/Redeemer01.plc:    1034
marlowe/trustfund/Redeemer02.plc:     610
marlowe/trustfund/Validator01.plc: 100707

marlowe/zerocoupon/Context01.plc:     3306
marlowe/zerocoupon/Context02.plc:     2305
marlowe/zerocoupon/Datum01.plc:       1642
marlowe/zerocoupon/Datum02.plc:        888
marlowe/zerocoupon/Redeemer01.plc:     857
marlowe/zerocoupon/Redeemer02.plc:     857
marlowe/zerocoupon/Validator01.plc: 100707
   
multisigSM/Context01.plc:    2071
multisigSM/Context02.plc:    2124
multisigSM/Context03.plc:    2177
multisigSM/Context04.plc:    2230
multisigSM/Context05.plc:    2106
multisigSM/Context06.plc:    2071
multisigSM/Context07.plc:    2124
multisigSM/Context08.plc:    2177
multisigSM/Context09.plc:    2230
multisigSM/Context10.plc:    2106
multisigSM/Datum01.plc:       333
multisigSM/Datum02.plc:       604
multisigSM/Datum03.plc:       657
multisigSM/Datum04.plc:       710
multisigSM/Datum05.plc:       763
multisigSM/Redeemer01.plc:    585
multisigSM/Redeemer02.plc:    386
multisigSM/Redeemer03.plc:    386
multisigSM/Redeemer04.plc:    386
multisigSM/Redeemer05.plc:    333
multisigSM/Validator01.plc: 38626

vesting/Context01.plc:    2108
vesting/Context02.plc:    1662
vesting/Datum01.plc:       333
vesting/Redeemer01.plc:    333
vesting/Validator01.plc: 21381
vesting/Validator02.plc: 21381
vesting/Validator03.plc: 21381
```

Reply:
Thanks. So they look relatively small compared to the validator scripts. Are those sizes in bytes?

Reply:
Actually those figures are wrong.  The numbers next to the `Datum/Redeeemer/Context` objects are sizes in bytes, but not of serialised `Data`.  In fact they're the sizes of CBOR-encoded Scott encodings of the data objects, ie after they've been deserialised and converted to PLC terms, de-Bruijnified, and then serialised again.  Presumably the serialised `Data` things are significantly smaller. I'll try to find out the proper answer when I have some spare time, and also how it compares with the previous scheme.